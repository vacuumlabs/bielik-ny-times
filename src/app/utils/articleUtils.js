import { parseArticleImage } from '../model/Image'

const sortByWidthAsc = (a, b) => a.width - b.width

/**
 * Searches for an article image, with appropriate size.
 * @param {ViewedArticle} articleResponse article response object
 * @see https://developer.nytimes.com/docs/most-popular-product/1/types/ViewedArticle
 */
export function findArticleImage(articleResponse = {}, { minWidth = 0 } = {}) {
  if (!articleResponse) {
    return null
  }
  const { media = [] } = articleResponse
  const firstImage = media.find(m => m.type === 'image')
  if (!firstImage) {
    return null
  }
  const images = firstImage['media-metadata'].slice()
  images.sort(sortByWidthAsc)
  const bestImage = images.find(image => image.width >= minWidth)
  if (bestImage) {
    return parseArticleImage(bestImage)
  }
  const largestImage = images[images.length - 1]
  return parseArticleImage(largestImage)
}
