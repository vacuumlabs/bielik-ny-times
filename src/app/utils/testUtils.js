import renderer from 'react-test-renderer'

export const renderToJSON = elem => renderer.create(elem).toJSON()

export const assertSnapshot = elem =>
  expect(renderToJSON(elem)).toMatchSnapshot()
