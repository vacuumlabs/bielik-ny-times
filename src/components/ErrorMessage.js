import { StyleSheet, Text } from 'react-native'
import React from 'react'

import { isApplicationError } from '../app/model/ApplicationError'
import theme from '../app/theme'

const TextError = ({ children }) => (
  <Text style={styles.error}>Error: {children}</Text>
)

/**
 * Component that can display various types of errors
 */
export default function ErrorMessage({
  error, // ApplicationError
}) {
  if (!error) {
    return null
  }
  if (isApplicationError(error)) {
    return <TextError>{error.message}</TextError>
  }
  return <TextError>Unexpected error occured :(</TextError>
}

const styles = StyleSheet.create({
  error: {
    color: theme.colors.danger,
  },
})
