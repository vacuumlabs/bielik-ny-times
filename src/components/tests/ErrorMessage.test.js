import React from 'react'

import { assertSnapshot } from '../../app/utils/testUtils'
import ErrorMessage from '../ErrorMessage'

describe('ErrorMessage', () => {
  it('should not render if missing', () =>
    assertSnapshot(<ErrorMessage error={undefined} />))

  it('should render application error', () =>
    assertSnapshot(<ErrorMessage error={{ message: 'app error' }} />))

  it('should render unknown error', () =>
    assertSnapshot(<ErrorMessage error={{ msg: 'something' }} />))
})
