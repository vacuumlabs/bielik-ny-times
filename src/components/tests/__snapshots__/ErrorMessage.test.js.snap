// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorMessage should not render if missing 1`] = `null`;

exports[`ErrorMessage should render application error 1`] = `
<Text
  style={
    Object {
      "color": "red",
    }
  }
>
  Error: 
  app error
</Text>
`;

exports[`ErrorMessage should render unknown error 1`] = `
<Text
  style={
    Object {
      "color": "red",
    }
  }
>
  Error: 
  Unexpected error occured :(
</Text>
`;
